sympolyc/src/expr.rs

```diff
@@ -0,0 +1,241 @@
+            }
+        });
+        normalized_indices
+    }
+
+    /// Returns `true` if the polynomial is homoegenous of degree `d`
+    pub fn is_homogeneous(&self) -> bool {
+        let normalized_indices = self.normalized_indices.clone();
+        let mut prime_gen = PrimeNumberGenerator::new();
+        let is_homogeneous = normalized_indices
+            .iter()
+            .zip(self.coeff.clone())
+            .all(|(idx, c)| {
+                let decomposition_of_i = naive_prime_factors(*idx, &mut prime_gen);
+                let monomial_degree = decomposition_of_i.iter().fold(0, |acc, (_, d)| acc + d);
+                !(monomial_degree != D && c != F::zero())
+            });
+        is_homogeneous
+    }
+}
+
+impl<F: PrimeField, const N: usize, const D: usize> Default for MVPoly<F, N, D> {
+    fn default() -> Self {
+        MVPoly::new()
+    }
+}
+
+// Addition
+impl<F: PrimeField, const N: usize, const D: usize> Add for MVPoly<F, N, D> {
+    type Output = Self;
+
+    fn add(self, other: Self) -> Self {
+        let mut result = MVPoly::new();
+        for i in 0..self.coeff.len() {
+            result.coeff[i] = self.coeff[i] + other.coeff[i];
+        }
+        result
+    }
+}
+
+impl<F: PrimeField, const N: usize, const D: usize> Add<&MVPoly<F, N, D>> for MVPoly<F, N, D> {
+    type Output = MVPoly<F, N, D>;
+
+    fn add(self, other: &MVPoly<F, N, D>) -> MVPoly<F, N, D> {
+        let mut result = MVPoly::new();
+        for i in 0..self.coeff.len() {
+            result.coeff[i] = self.coeff[i] + other.coeff[i];
+        }
+        result
+    }
+}
+
+impl<F: PrimeField, const N: usize, const D: usize> Add<MVPoly<F, N, D>> for &MVPoly<F, N, D> {
+    type Output = MVPoly<F, N, D>;
+
+    fn add(self, other: MVPoly<F, N, D>) -> MVPoly<F, N, D> {
+        let mut result = MVPoly::new();
+        for i in 0..self.coeff.len() {
+            result.coeff[i] = self.coeff[i] + other.coeff[i];
+        }
+        result
+    }
+}
+
+impl<F: PrimeField, const N: usize, const D: usize> Add<&MVPoly<F, N, D>> for &MVPoly<F, N, D> {
+    type Output = MVPoly<F, N, D>;
+
+    fn add(self, other: &MVPoly<F, N, D>) -> MVPoly<F, N, D> {
+        let mut result = MVPoly::new();
+        for i in 0..self.coeff.len() {
+            result.coeff[i] = self.coeff[i] + other.coeff[i];
+        }
+        result
+    }
+}
+
+// Multiplication
+impl<F: PrimeField, const N: usize, const D: usize> Mul<MVPoly<F, N, D>> for MVPoly<F, N, D> {
+    type Output = Self;
+
+    fn mul(self, other: Self) -> Self {
+        let mut acc = HashMap::new();
+        let mut prime_gen = PrimeNumberGenerator::new();
+        let mut result = vec![];
+        (0..self.coeff.len()).for_each(|i| {
+            let mut sum = F::zero();
+            let normalized_index = self.normalized_indices[i];
+            let two_factors_decomposition =
+                compute_all_two_factors_decomposition(normalized_index, &mut acc, &mut prime_gen);
```

Comment:
This is a weird way to iterate, or is it just me ?
You are finding all the possible product of monomials that can contribute to a coeff of the result.
I would do the schoolbook way, which iterates on both polynomials, and compute which contribution goes where.

Reply:
It is the whole point of this approach using prime numbers.